src/Microsoft.Data.SqlClient/add-ons/AzureKeyVaultProvider/AzureKeyVaultProviderTokenCredential.cs

```diff
@@ -0,0 +1,90 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net.Http;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Azure.Core;
+
+namespace Microsoft.Data.SqlClient.AlwaysEncrypted.AzureKeyVaultProvider
+{
+    internal class AzureKeyVaultProviderTokenCredential : TokenCredential
+    {
+        private const string Bearer = "Bearer ";
+
+        private AuthenticationCallback Callback { get; set; }
+
+        private string Authority { get; set; }
+
+        private string Resource { get; set; }
+
+        internal AzureKeyVaultProviderTokenCredential(AuthenticationCallback authenticationCallback, string masterKeyPath)
+        {
+            Callback = authenticationCallback;
+            using (HttpClient httpClient = new HttpClient())
+            {
+                HttpResponseMessage response = httpClient.GetAsync(masterKeyPath).GetAwaiter().GetResult();
+                string challenge = response?.Headers.WwwAuthenticate.FirstOrDefault()?.ToString();
+                string trimmedChallenge = ValidateChallenge(challenge);
+                string[] pairs = trimmedChallenge.Split(new string[] { "," }, StringSplitOptions.RemoveEmptyEntries);
+
+                if (pairs != null && pairs.Length > 0)
+                {
+                    for (int i = 0; i < pairs.Length; i++)
+                    {
+                        string[] pair = pairs[i]?.Split('=');
+
+                        if (pair.Length == 2)
+                        {
+                            string key = pair[0]?.Trim().Trim(new char[] { '\"' });
+                            string value = pair[1]?.Trim().Trim(new char[] { '\"' });
+
+                            if (!string.IsNullOrEmpty(key))
+                            {
+                                if (key.Equals("authorization", StringComparison.InvariantCultureIgnoreCase))
+                                {
+                                    Authority = value;
+                                }
+                                else if (key.Equals("resource", StringComparison.InvariantCultureIgnoreCase))
+                                {
+                                    Resource = value;
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        public override AccessToken GetToken(TokenRequestContext requestContext, CancellationToken cancellationToken)
+        {
+            string token = Callback.Invoke(Authority, Resource, string.Empty).GetAwaiter().GetResult();
+            return new AccessToken(token, DateTimeOffset.Now.AddHours(1));
```

Comment:
See https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/keyvault/Azure.Security.KeyVault.Shared/src/ChallengeBasedAuthenticationPolicy.cs#L112 for how we handle expiration.

Reply:
In our case token is just the "string" access token, outcome of below, we don't have "expirationTime" from callback.

```
public static async Task<string> AzureActiveDirectoryAuthenticationCallback(string authority, string resource, string scope)
{
    var authContext = new AuthenticationContext(authority);
    ClientCredential clientCred = new ClientCredential(s_clientId, s_clientSecret);
    AuthenticationResult result = await authContext.AcquireTokenAsync(resource, clientCred);
    if (result == null)
    {
        throw new InvalidOperationException($"Failed to retrieve an access token for {resource}");
    }
    return result.AccessToken;
}
```

Reply:
Per internal discussion, you may be able to write a pipeline policy to get the authority from the challenge response (rather, the request back to authenticate the challenge.) and pass that to the callback. You'd have to associate that with the request - perhaps stored as a field in your token credential, but in a way where if the token credential is used for multiple clients, is associated with the right  resource (vault URI).